File: semaforos/exemploLivro/signal/signaling31.py
import threading

semaphore = threading.Semaphore(0)

def thread_function_a():
    print("A Chegou\n")
    semaphore.release()

def thread_function_b():
    print("Thread B: Esperando pelo sinal de A...\n")
    semaphore.acquire()
    print("B Chegou")

if __name__ == '__main__':
    t1 = threading.Thread(target=thread_function_a, name="A")
    t2 = threading.Thread(target=thread_function_b, name="B")

    t2.start()
    t1.start()
    t1.join()
    t2.join()
    

File: semaforos/exemploLivro/signal/signaling31.java
package semaforos.exemploLivro.signal;

import java.util.concurrent.Semaphore;

public class signaling31 implements Runnable{
    private Semaphore aArrive = new Semaphore(0);

    public static void main(String[] args) {
        Thread t1 = new Thread(new signaling31(), "A");
        Thread t2 = new Thread(new signaling31(), "b");
        t2.start();
        t1.start();
    }

    @Override
    public void run() {

        String name = Thread.currentThread().getName();

        switch (name) {
            case "A":
                System.out.println("A Chegou");
                aArrive.release();
                break;
            case "B":
            try {
                aArrive.acquire();
            } catch (InterruptedException e) {
                System.err.println(e.getStackTrace());
            }
            default:
                break;
        }

    }
}
